formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Decoder.kt

```diff
@@ -0,0 +1,636 @@
+            throw CborDecodingException("start of string", curByte)
+        val arr = readBytes()
+        val ans = arr.decodeToString()
+        readByte()
+        return ans to collectedTags
+    }
+
+    private fun readBytes(): ByteArray =
+        if (curByte and 0b000_11111 == ADDITIONAL_INFORMATION_INDEFINITE_LENGTH) {
+            readByte()
+            readIndefiniteLengthBytes()
+        } else {
+            val strLen = readNumber().toInt()
+            input.readExactNBytes(strLen)
+        }
+
+    @SuppressAnimalSniffer
+    private fun processTags(tags: ULongArray?): ULongArray? {
+        var index = 0
+        val collectedTags = mutableListOf<ULong>()
+        while ((curByte and 0b111_00000) == HEADER_TAG) {
+            val readTag = readNumber().toULong() // This is the tag number
+            collectedTags += readTag
+            // value tags and object tags are intermingled (keyTags are always separate)
+            // so this check only holds if we verify both
+            if (verifyObjectTags) {
+                tags?.let {
+                    if (index++ > it.size) throw CborDecodingException("More tags found than the ${it.size} tags specified")
+                    if (readTag != it[index - 1]) throw CborDecodingException("CBOR tag $readTag does not match expected tag $it")
+                }
+            }
+            readByte()
+        }
+        return (if (collectedTags.isEmpty()) null else collectedTags.toULongArray()).also { collected ->
+            //We only want to compare if tags are actually set, otherwise, we don't care
+            tags?.let {
+                if (verifyObjectTags) { //again, this check only works if we verify value tags and object tags
+                    if (!it.contentEquals(collected))
+                        throw CborDecodingException(
+                            "CBOR tags ${
+                                collected?.joinToString(
+                                    prefix = "[",
+                                    postfix = "]"
+                                ) { it.toString() }
+                            } do not match expected tags ${
+                                it.joinToString(
+                                    prefix = "[",
+                                    postfix = "]"
+                                ) { it.toString() }
+                            }"
+                        )
+                } else {
+                    // If we don't care for object tags, the best we can do is assure that the collected tags start with
+                    // the expected tags. (yes this could co somewhere else, but putting it here groups the code nicely
+                    // into if-else branches.
+                    if ((collectedTags.size < it.size)
+                        || (collectedTags.subList(0, it.size) != it.asList())
+                    ) throw CborDecodingException("CBOR tags $collectedTags does not start with specified tags $it")
```

Comment:
```suggestion
                    ) throw CborDecodingException("CBOR tags $collectedTags do not start with specified tags $it")
```

Reply:
You are right, the second check in the while loop can never be triggered, the idea was an early exit for obviously wrong tags, but that's not how it works, so the second check is gone in 6e19790669fbe1c78992e0353fed2a6d2882f6c2.

More context: 
In the while loop, we're essentially collecting (i.e. consuming) all tags. If we verify and encounter too many tags, we exit early. Consumption of tags needs to happen anyhow, regardless of whether we verify tags or not. This check here is the actual check against the tags from the annotations. We only want to perform this check, if the tags annotation is present (i.e. if tags are actually passed).

This affair is tricker than testing for KeyTags, due to ObjectTags being a thing.